foundation/provider_repository.go

```diff
@@ -0,0 +1,377 @@
+package foundation
+
+import (
+	"fmt"
+	"sort"
+	"strings"
+
+	"github.com/goravel/framework/contracts/binding"
+	"github.com/goravel/framework/contracts/foundation"
+	"github.com/goravel/framework/errors"
+)
+
+type ProviderState struct {
+	provider   foundation.ServiceProvider
+	registered bool
+	booted     bool
+}
+
+var _ foundation.ProviderRepository = (*ProviderRepository)(nil)
+
+type ProviderRepository struct {
+	states      map[string]*ProviderState
+	providers   []foundation.ServiceProvider
+	sorted      []foundation.ServiceProvider
+	sortedValid bool
+	loaded      bool
+}
+
+func NewProviderRepository() *ProviderRepository {
+	return &ProviderRepository{
+		states:    make(map[string]*ProviderState),
+		providers: make([]foundation.ServiceProvider, 0),
+	}
+}
+
+func (r *ProviderRepository) Register(app foundation.Application) []foundation.ServiceProvider {
```

Comment:
`app foundation.Application` can be removed, given it has been passed in `NewProviderRepository`.

```suggestion
func (r *ProviderRepository) Register() []foundation.ServiceProvider {
```